In [101]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression




---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [102]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
file_path = Path("Resources/lending_data.csv")
lending_df = pd.read_csv(file_path)

# Review the DataFrame
# YOUR CODE HERE!
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [103]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = lending_df['loan_status']

# Separate the X variable, the features
# YOUR CODE HERE!
X = lending_df.drop(columns='loan_status')

In [104]:
# Review the y variable Series
y.sample(5)

47827    0
76025    1
15718    0
22300    0
32699    0
Name: loan_status, dtype: int64

In [105]:
# Review the X variable DataFrame
X.sample(5)

loan_size  interest_rate  borrower_income  debt_to_income  \
47357    10400.0          7.555            51700        0.419729   
75904    19000.0         11.186            85900        0.650757   
25647     9300.0          7.098            47400        0.367089   
72766     9600.0          7.197            48300        0.378882   
18006    10100.0          7.402            50300        0.403579   

       num_of_accounts  derogatory_marks  total_debt  
47357                4                 1       21700  
75904               12                 2       55900  
25647                3                 0       17400  
72766                4                 0       18300  
18006                4                 1       20300

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [106]:
# Check the balance of our target values
# YOUR CODE HERE!
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [107]:
# Import the train_test_learn module


# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [108]:
# Import the LogisticRegression module from SKLearn


# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
classifier = LogisticRegression(random_state = 1)

# Fit the model using training data
# YOUR CODE HERE!
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [109]:
# Make a prediction using the testing data
# YOUR CODE HERE!
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test}).tail(10)


Prediction  Actual
76038           1       1
19119           0       0
45193           0       0
25211           0       0
61697           0       0
38069           0       0
36892           0       0
5035            0       0
40821           0       0
35030           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [110]:
# Print the balanced_accuracy_score of the model
# YOUR CODE HERE!
# classifier.score we learned followed by the balanced_accuracy_score we did not learn in class.
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test, predictions)}")

Training Data Score: 0.9914878250103177
Testing Data Score: 0.9924164259182832
Balanced Accuracy Score: 0.9442676901753825


In [111]:
# Generate a confusion matrix for the model
# Out of The model predicted 18679 loans correctly and 80 incorrectly for the healthy loans.
# The model prdicted 558 loans correctly and 67 incorrectly for the non-healthy loans.

# YOUR CODE HERE!

c_matrix = confusion_matrix(y_test, predictions)
confusion_matrix_df = pd.DataFrame(c_matrix,
                                   index=["Healthy Loans","Non-Healthy Loans"],
                                   columns=["Predicted Healthy Loans", "Predicted Non-Healthy Loans"])
confusion_matrix_df

Predicted Healthy Loans  Predicted Non-Healthy Loans
Healthy Loans                        18679                           80
Non-Healthy Loans                       67                          558

In [112]:
# Print the classification report for the model
# YOUR CODE HERE!

print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** According to the classification report the model predicted correctly: healthy loans (0) 100% of the time and high-risk loans (1) 87% of the time.
The Balanced Accuracy Score is 0.9442676901753825.


---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [113]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
ROS_model = RandomOverSampler(random_state=1)
ROS_model
# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
ROS_model.fit(X_train, y_train)
X_over, y_over = ROS_model.fit_resample(X_train, y_train)


In [114]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
y_over.value_counts()


0    56277
1    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [115]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
LR_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
# YOUR CODE HERE!
LR_model.fit(X_over, y_over)

# Make a prediction using the testing data
# YOUR CODE HERE!
LR_prediction = LR_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [116]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test, LR_prediction)}")

Balanced Accuracy Score: 0.9959744975744975


In [117]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
c1_matrix = confusion_matrix(y_test, LR_prediction)
confusion1_matrix_df = pd.DataFrame(c1_matrix,
                                   index=["Healthy Loans","Non-Healthy Loans"],
                                   columns=["Predicted Healthy Loans", "Predicted Non-Healthy Loans"])
confusion1_matrix_df


Predicted Healthy Loans  Predicted Non-Healthy Loans
Healthy Loans                        18668                           91
Non-Healthy Loans                        2                          623

In [118]:
# Print the classification report for the model
# YOUR CODE HERE!
print(classification_report(y_test, LR_prediction))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      1.00      0.93       625

    accuracy                           1.00     19384
   macro avg       0.94      1.00      0.96     19384
weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** According to the classification report the model predicted correctly: healthy loans (0) 100% of the time and high-risk loans (1) 87% of the time.
The Balanced Accuracy Score: 0.9959744975744975!!!! This has improved to a very high rate of success.